Refactor recompress_chunk() API

sb230132 opened the pull request:
This patch will improve performance of recompress_chunk() API by identifying only affected segments which needs to be recompressed. All unaffected segments are never decompressed.

mkindahl commented on tsl/src/compression/api.c:
```diff
@@ -1075,128 +1059,682 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+						   HeapTuple *compressed_tuple)
+{
+	/* build scankeys for segmentby columns */
+	ScanKeyData *scankeys = palloc0(nsegmentby_cols * sizeof(ScanKeyData));
```
Not strictly necessary, but maybe you want to release the memory since you know it will not be needed after the scan.

sb230132 replied:
Fixed.

mkindahl commented on tsl/src/compression/api.c:
```diff
@@ -1075,128 +1059,682 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+{
+	int segments_decompressed = 0;
+	/* build scankeys for segmentby and sequence number columns */
+	ScanKeyData *scankeys = palloc0((nsegmentby_cols + 1) * sizeof(ScanKeyData));
```
You might want to release this memory after you're done with it. Doesn't seem like it need to remain after the scan.

sb230132 replied:
Fixed.

sb230132 commented on tsl/test/isolation/specs/compression_ddl_iso.spec:
```diff
@@ -44,7 +44,7 @@ session "S"
 step "SC1" { SELECT (count_chunktable(ch)).* FROM show_chunks('ts_device_table') AS ch ORDER BY ch::text LIMIT 1; }
 step "SH" { SELECT total_chunks, number_compressed_chunks from hypertable_compression_stats('ts_device_table'); }
-step "SA" { SELECT * FROM ts_device_table; }
+step "SA" { SELECT * FROM ts_device_table ORDER BY 1; }
```
Fixed.

konskov commented on tsl/test/sql/CMakeLists.txt:
```diff
@@ -114,7 +114,7 @@ if((${PG_VERSION_MAJOR} GREATER_EQUAL "14"))
   endif()
   list(APPEND TEST_FILES compression.sql compression_update_delete.sql
-       compression_permissions.sql)
+       compression_permissions.sql recompress_chunk_segmentwise-v2.sql)
```
really minor, just thought it might be useful to add an explanation why this is only tested for pg14 and later.

sb230132 replied:
Fixed, added comment.
There is no Custom Scan (HypertableModify) node for UPDATE/DELETE operations on PG versions < 14, and to support UPDATE/DELETE  operations on compressed chunks we need HypertableModify to be part of plan.
Thus we don't support this feature on PG12 and PG13.

konskov commented on tsl/src/compression/compression.h:
```diff
@@ -260,6 +261,7 @@ typedef struct RowCompressor
 	int64 num_compressed_rows;
 	/* if recompressing segmentwise, we use this info to reset the sequence number */
 	bool reset_sequence;
+	bool update_sequence;
```
minor, might be useful to add a comment explaining when this field is used?

sb230132 replied:
Fixed, added comment.

konskov commented on tsl/test/sql/recompress_chunk_segmentwise-v2.sql:
```diff
@@ -0,0 +1,477 @@
+
+\c :TEST_DBNAME :ROLE_SUPERUSER
+DROP DATABASE IF EXISTS recompression_optimization_db;
+CREATE DATABASE recompression_optimization_db;
```
sorry if it's obvious but why do we need to create a new database for this test?

sb230132 replied:
Right, removed it now.

konskov commented on tsl/src/compression/api.c:
```diff
@@ -1075,128 +1061,692 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+													 sequence_num,
+													 false);
+		HeapTuple compressed_tuple;
+		TableScanDesc heapScan =
```
I was wondering if it might be possible/more beneficial to do an index scan here on the compressed chunk instead of a heapscan?

sb230132 replied:
On a compressed chunk, we have an index defined on segmentby column and sequence number. Thus as you say i can use the index to scan the chunk.
Made necessary changes too.

konskov commented on tsl/src/compression/api.c:
```diff
@@ -1075,128 +1061,692 @@ tsl_get_compressed_chunk_index_for_recompression(PG_FUNCTION_ARGS)
+																uncompressed_slot,
+																&num_scankeys);
+
+		TableScanDesc heapScan = table_beginscan(decompressor.in_rel,
```
(same thing here, maybe doing an indexscan instead of heapscan on the compressed chunk)

sb230132 replied:
At this place, its not possible to use index scan. I need to build scankeys on segmentby and orderby columns. There is no index defined on compressed chunk which has order by columns.
So keeping the code as is.

akuzm reviewed (changes requested):
If we look at the greater picture, there are several ways we make use of having some granularity in the compressed data. By granularity I mean that we use small compressed batches and not compress the entire chunk in one blob.

- Read granularity. We don't want to decompress the entire chunk into memory if we only need to read a few rows.
- Indexing. We can find the matching compressed batches by their metadata, to reduce the amount of reads.
- Updates. We can decompress only matching batches and not the entire chunk, when one row is updated.
- Simplified decompression. When there is an upper limit on the size of a compressed batch, we can write a simpler and more efficient bulk algorithm for decompression, instead of using a streaming algorithm to work with potentially ulimited data.
- Generally, knowing that one compressed batch fits into a small amount of memory. We use this assumption in DecompressChunk node, even w/o the bulk decompression.

So these are a few reasons why we limit the compressed batch size, and there are probably more. We use 1000 rows, another DB I know uses 8129.

If I understand things right, this PR tries to lift this limit, and replace it with... what exactly? From a brief look at the code, it seems to me that there would be no limit at all. I don't think it is feasible to work with unlimited-size compressed chunks, because of the aforementioned issues.

sb230132 commented:
> If we look at the greater picture, there are several ways we make use of having some granularity in the compressed data. By granularity I mean that we use small compressed batches and not compress the entire chunk in one blob.
> 
> * Read granularity. We don't want to decompress the entire chunk into memory if we only need to read a few rows.
> * Indexing. We can find the matching compressed batches by their metadata, to reduce the amount of reads.
> * Updates. We can decompress only matching batches and not the entire chunk, when one row is updated.
> * Simplified decompression. When there is an upper limit on the size of a compressed batch, we can write a simpler and more efficient bulk algorithm for decompression, instead of using a streaming algorithm to work with potentially ulimited data.
> * Generally, knowing that one compressed batch fits into a small amount of memory. We use this assumption in DecompressChunk node, even w/o the bulk decompression.
> 
> So these are a few reasons why we limit the compressed batch size, and there are probably more. We use 1000 rows, another DB I know uses 8129.
> 
> If I understand things right, this PR tries to lift this limit, and replace it with... what exactly? From a brief look at the code, it seems to me that there would be no limit at all. I don't think it is feasible to work with unlimited-size compressed chunks, because of the aforementioned issues.

Ok, i dint know the actual reason behind having a limit of 1000 per segment. Thanks for the detailed explaination.
Renamed the guc to "increase_max_rows_per_compression" which when enabled will allow a max of upto 2000 rows per segment.

akuzm commented:
> Renamed the guc to "increase_max_rows_per_compression" which when enabled will allow a max of up to 2000 rows per segment during recompression.

Can we remove this part from the current PR at all? I think changing the batch size will need some additional benchmarking. Also this would have complicated performance characteristics -- i.e. you can insert twice as much data and it's fast, but then you insert more and suddenly it hits a different path and becomes slow. We are going to get SDCs about this :)

Also I was wondering, when exactly is this optimization helpful? If we can append the new row to the end of last batch w/o breaking the "order by", it also means we can start a new batch with higher sequence number, right?

sb230132 commented:
> > Renamed the guc to "increase_max_rows_per_compression" which when enabled will allow a max of up to 2000 rows per segment during recompression.
> 
> Can we remove this part from the current PR at all? I think changing the batch size will need some additional benchmarking. Also this would have complicated performance characteristics -- i.e. you can insert twice as much data and it's fast, but then you insert more and suddenly it hits a different path and becomes slow. We are going to get SDCs about this :)
> 
> Also I was wondering, when exactly is this optimization helpful? If we can append the new row to the end of last batch w/o breaking the "order by", it also means we can start a new batch with higher sequence number, right?

I don't want to change the default value of MAX_ROWS_PER_COMPRESSION (1000). As you say this requires more amount of performance/benchmarking testing, also requires approval from product.

All i do is introduce a new guc variable which when enabled, will increase MAX_ROWS_PER_COMPRESSION to 2000 only for recompression operations.

Please check below how the optimisation works. If you notice number of segments decompressed is 1 (with optimisation) vs 4   (without optimisation).

```
-- case1: 1 segment affected
-- Number of segments decompressed is 1
BEGIN;
-- affects segment with sequence_num = 10
INSERT INTO metrics_compressed VALUES ('2000-01-05 20:06:00+05:30'::timestamp with time zone, 1, 2, 3, 22.2, 33.3);
INSERT INTO metrics_compressed VALUES ('2000-01-04 23:06:00+05:30'::timestamp with time zone, 1, 2, 3, 222.2, 333.3);
-- check compression status
SELECT * FROM _timescaledb_catalog.chunk WHERE table_name = '_hyper_1_1_chunk' ORDER BY id;
 id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status | osm_chunk 
----+---------------+-----------------------+------------------+---------------------+---------+--------+-----------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                   6 | f       |      9 | f
(1 row)

SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |           1000 | Tue Jan 04 06:36:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:16:00 2000 PST | Tue Jan 04 06:34:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 11:56:00 2000 PST | Sun Jan 02 21:14:00 2000 PST
         1 |                    40 |            598 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:54:00 2000 PST
(4 rows)

SET timescaledb.increase_max_rows_per_compression = ON;
SET client_min_messages TO LOG;
CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  statement: CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  Total decompressed segments are: 1
RESET client_min_messages;
LOG:  statement: RESET client_min_messages;
SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |           1002 | Tue Jan 04 06:36:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:16:00 2000 PST | Tue Jan 04 06:34:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 11:56:00 2000 PST | Sun Jan 02 21:14:00 2000 PST
         1 |                    40 |            598 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:54:00 2000 PST
(4 rows)

-- check compression status
SELECT * FROM _timescaledb_catalog.chunk WHERE table_name = '_hyper_1_1_chunk' ORDER BY id;
 id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status | osm_chunk 
----+---------------+-----------------------+------------------+---------------------+---------+--------+-----------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                   6 | f       |      1 | f
(1 row)

ROLLBACK;
-- case1: 1 segment affected with increase_max_rows_per_compression = OFF
-- Number of segments decompressed is 4
BEGIN;
-- affects segment with sequence_num = 10
INSERT INTO metrics_compressed VALUES ('2000-01-05 20:06:00+05:30'::timestamp with time zone, 1, 2, 3, 22.2, 33.3);
INSERT INTO metrics_compressed VALUES ('2000-01-04 23:06:00+05:30'::timestamp with time zone, 1, 2, 3, 222.2, 333.3);
-- check compression status
SELECT * FROM _timescaledb_catalog.chunk WHERE table_name = '_hyper_1_1_chunk' ORDER BY id;
 id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status | osm_chunk 
----+---------------+-----------------------+------------------+---------------------+---------+--------+-----------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                   6 | f       |      9 | f
(1 row)

SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |           1000 | Tue Jan 04 06:36:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:16:00 2000 PST | Tue Jan 04 06:34:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 11:56:00 2000 PST | Sun Jan 02 21:14:00 2000 PST
         1 |                    40 |            598 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:54:00 2000 PST
(4 rows)

set client_min_messages TO LOG;
CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  statement: CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  Total decompressed segments are: 4
RESET client_min_messages;
LOG:  statement: RESET client_min_messages;
SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |           1000 | Tue Jan 04 06:40:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:20:00 2000 PST | Tue Jan 04 06:38:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 12:00:00 2000 PST | Sun Jan 02 21:18:00 2000 PST
         1 |                    40 |            600 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:58:00 2000 PST
(4 rows)
```

Hope i answered your questions.

akuzm commented:
>  I don't want to change the default value of MAX_ROWS_PER_COMPRESSION (1000). As you say this requires more amount of performance/benchmarking testing, also requires approval from product.

But this PR is changing it, and in a more complicated way: now it's not a constant, but depends on insert pattern. This requires even more complex testing than a new constant value. By the way, `GLOBAL_MAX_ROWS_PER_COMPRESSION` should also be changed to adapt bulk decompression to this.

> Please check below how the optimisation works. If you notice number of segments decompressed is 1 (with optimisation) vs 4 (without optimisation).

Thanks for the example. Instead of increasing the size of the first batch to 1002, is it possible to split it in two, with sequence numbers 10 and 11? Then we won't have to touch the next batches, and won't have to increase the max size.

sb230132 commented:
> But this PR is changing it, and in a more complicated way: now it's not a constant, but depends on insert pattern. This requires even more complex testing than a new constant value. By the way, `GLOBAL_MAX_ROWS_PER_COMPRESSION` should also be changed to adapt bulk decompression to this.

Oh yes, I see that compress_chunk() is also affected with my changes.
I get it now. What if initial few segments have max rows to 1000 and later we increase the threshold and we end up with segments having rows > 1000. Thus a chunk can have segments with a mix of max rows per segments (which can be < 1000, <= 1000, > 1000, <= 2000)

I also need to change GLOBAL_MAX_ROWS_PER_COMPRESSION accordingly.
What are you suggesting. Should we drop this. Or increase the constant of MAX_ROWS_PER_COMPRESSION to 2000 ?

> Thanks for the example. Instead of increasing the size of the first batch to 1002, is it possible to split it in two, with sequence numbers 10 and 11? Then we won't have to touch the next batches, and won't have to increase the max size.

Ok, this seems to be an alternative approach. Good idea !! So for a given segment we can have sequence numbers ranging from 10... 19.  What if segments with sequence numbers from 10 ... 19 gets full. Should we move rest of the rows to next segment with sequence number 20 and so on ?

Bear with me please, i want it to be clear before i can fix the implementation.

sb230132 commented:
> > What are you suggesting. Should we drop this. Or increase the constant of MAX_ROWS_PER_COMPRESSION to 2000 ?
> 
> I would drop it for now, because just increasing the constant won't solve the problem -- your optimization won't work then. Let's consider the alternative approach with sequence numbers. I think the purpose of gaps in sequence numbers (they increment by 10) is exactly this class of scenarios -- to allow adding more data later without repacking the entire table.
> 
> > So for a given segment we can have sequence numbers ranging from 10... 19. What if segments with sequence numbers from 10 ... 19 gets full. Should we move rest of the rows to next segment with sequence number 20 and so on ?
> 
> I think we can fall back to recompressing the subsequent batches. So if we have free sequence numbers before the next batch, we can insert the new one, but if we don't, we should recompress it as well.

I made changes as per your suggestions. Please have a look.
Below is a sample output where we split segments into 2 equal halfs.
```
BEGIN;
-- affects segment with sequence_num = 10
INSERT INTO metrics_compressed VALUES ('2000-01-05 20:06:00+05:30'::timestamp with time zone, 1, 2, 3, 22.2, 33.3);
INSERT INTO metrics_compressed VALUES ('2000-01-04 23:06:00+05:30'::timestamp with time zone, 1, 2, 3, 222.2, 333.3);
-- check compression status
SELECT * FROM _timescaledb_catalog.chunk WHERE table_name = '_hyper_1_1_chunk' ORDER BY id;
 id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status | osm_chunk 
----+---------------+-----------------------+------------------+---------------------+---------+--------+-----------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                   6 | f       |      9 | f
(1 row)

SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |           1000 | Tue Jan 04 06:36:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:16:00 2000 PST | Tue Jan 04 06:34:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 11:56:00 2000 PST | Sun Jan 02 21:14:00 2000 PST
         1 |                    40 |            598 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:54:00 2000 PST
(4 rows)

set client_min_messages TO LOG;
CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  statement: CALL recompress_chunk('_timescaledb_internal._hyper_1_1_chunk');
LOG:  Total decompressed segments are: 1
RESET client_min_messages;
LOG:  statement: RESET client_min_messages;
SELECT device_id, _ts_meta_sequence_num, _ts_meta_count, _ts_meta_min_1, _ts_meta_max_1 FROM _timescaledb_internal.compress_hyper_2_6_chunk WHERE device_id = 1;
 device_id | _ts_meta_sequence_num | _ts_meta_count |        _ts_meta_min_1        |        _ts_meta_max_1        
-----------+-----------------------+----------------+------------------------------+------------------------------
         1 |                    10 |            501 | Tue Jan 04 23:16:00 2000 PST | Wed Jan 05 15:54:00 2000 PST
         1 |                    11 |            501 | Tue Jan 04 06:36:00 2000 PST | Tue Jan 04 23:14:00 2000 PST
         1 |                    20 |           1000 | Sun Jan 02 21:16:00 2000 PST | Tue Jan 04 06:34:00 2000 PST
         1 |                    30 |           1000 | Sat Jan 01 11:56:00 2000 PST | Sun Jan 02 21:14:00 2000 PST
         1 |                    40 |            598 | Fri Dec 31 16:00:00 1999 PST | Sat Jan 01 11:54:00 2000 PST
(5 rows)

-- check compression status
SELECT * FROM _timescaledb_catalog.chunk WHERE table_name = '_hyper_1_1_chunk' ORDER BY id;
 id | hypertable_id |      schema_name      |    table_name    | compressed_chunk_id | dropped | status | osm_chunk 
----+---------------+-----------------------+------------------+---------------------+---------+--------+-----------
  1 |             1 | _timescaledb_internal | _hyper_1_1_chunk |                   6 | f       |      1 | f
(1 row)

ROLLBACK;
```

sb230132 commented:
In short this patch does not perform well if all segments in compressed chunks needs to be recompressed.
![image](https://github.com/timescale/timescaledb/assets/6995422/b2a8421c-08e6-4811-9582-fe2e8b3e0971)

akuzm commented on tsl/test/expected/recompress_chunk_segmentwise.out:
```diff
@@ -146,18 +146,18 @@ select _timescaledb_internal.recompress_chunk_segmentwise(:'chunk_to_compress_2'
 (2 rows)
 
 -- verify that initial data is returned as expected
 select * from :chunk_to_compress_2;
```
Let's add `ORDER BY` to such queries to prevent flaky output.

sb230132 replied:
Fixed.

akuzm commented:
> This new patch performs well if atmost 50% of segments are affected. Anything more that this will cause a degradation in performance.

Is there a deep reason for it? Maybe it's possible for the new recompression to do as well as the old one? After all, it starts from the same data and produces about the same result, if I understand it correctly.

akuzm commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,863 @@
+	if ((meta_count + uncompressed_rows_per_segment) >= MAX_ROWS_PER_COMPRESSION)
+		return (meta_count + uncompressed_rows_per_segment);
+	else
+		return (2 * MAX_ROWS_PER_COMPRESSION);
```
Can we just avoid this kind of arIthmetics at all? If we have such code, we have to think hard to prove that the actual limit is still `MAX_ROWS_PER_COMPRESSION`.


I can imagine roughly the following way to deal with overflowing batches:

We have to incorporate new rows into the already compressed data. To understand what work is required, we will track the ranges of compressed data that are affected by the recompression. Compressed data is ordered by (segmentby, orderby) values. One range is an interval on this axis and the respective metadata. 

First, go over the uncompressed rows ordered by (segmentby, orderby).
Keep track of the current affected compressed range:
	- segmentby
	- min (orderby)
	- max (orderby)
	- row count
	- starting sequence number
	- next adjacent sequence number (of the first next batch that is not affected)

For each row:
    - does it fall into the current range?
    	* if not, the current range is ready for compression! remember it and start the new one with this row.

	- find the intersecting compressed batch
		* if found, adjust the range to include it

	- check that the current compressed range is re-compressible by itself, without affecting the compressed batches: it should have enough sequence numbers to fit its count of rows as compressed batches of max size
		* if not, include the next compressed segment into the range
		* loop until true

After we built these compressed ranges, we know that each can be compressed using our normal algorithm, and we will have enough sequence numbers for it.
 
I don't like that we resort after each compressed segment, this is a lot of extra work, would be good to find a way to only re-sort once. Maybe we could put the decompressed batches into a second tuplesort.

Do you think we could refactor it along these lines?

akuzm commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,863 @@
+fetch_segment_metadata_count(RowDecompressor *decompressor, HeapTuple compressed_tuple,
+							 Datum *metacount_val)
+{
+	Datum *values = palloc0(sizeof(Datum) * decompressor->in_desc->natts);
+	bool *nulls = palloc0(sizeof(bool) * decompressor->in_desc->natts);
+	heap_deform_tuple(compressed_tuple, decompressor->in_desc, values, nulls);
```
`slot_getattr` will do that for you, and in a more efficient way (not all columns but only up to the attribute that is required, this is faster in Postgres).

sb230132 commented:
Implemented new patch as suggested.
Introduced a new data structure called ``` AffectedSegmentsCxt ```
This structure will have a mapping between uncompressed rows and compressed rows matching the segmentby values.
For all uncompressed rows with new  segmentby values there will be no corresponding compressed rows .

This structure is populated in ``` save_affected_compressed_rows_using_indexscan() ``` and ``` save_affected_compressed_rows_using_heapscan() ``` based on the presence of an index on compressed chunk.
Due to several different corner cases which needs to be handled these methods does not seem so easy to understand. 
All i can do is to add comment to the code as much as possible.

The consumers of this new data structure is: ``` recompress_affected_segments() ``` and ``` recompress_new_segments() ``` which will recompress all the compressed/uncompressed rows.

With this new patch the performance has been greatly improved.
![image](https://github.com/timescale/timescaledb/assets/6995422/2ccaafbd-48b7-4cad-a22a-9e7a7baa879c)

Please refer: https://docs.google.com/spreadsheets/d/1A1OmCfD3mP_cTNqar9YQ20T9puY6xAoLJColYRKKCIs/edit?usp=sharing

antekresic commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1520 @@
+		ExecStoreHeapTuple(uncompressed_tuple, heap_tuple_slot, false);
+		slot_getallattrs(heap_tuple_slot);
+		tuplesort_puttupleslot(uncompressed_rows_sortstate, heap_tuple_slot);
+		simple_heap_delete(uncompressed_chunk_rel, &uncompressed_tuple->t_self);
```
Could we not truncate the relation?

sb230132 replied:
Fixed, thanks.

antekresic reviewed (commented):
First pass, we do another one.

Can we add another tests with index removed on compressed chunk to verify heap scan is working correctly?

sb230132 commented:
> First pass, we do another one.
> 
> Can we add another tests with index removed on compressed chunk to verify heap scan is working correctly?

Added testcase for heapscan, by dropping the index on compressed chunk.

erimatnor commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1449 @@
+				 * we cannot use HeapKeyTest to check for NULLS, thus perform
+				 * manual checks by calling has_all_nulls()
+				 */
+				uncompressed_tuple = heap_form_tuple(uncompressed_slot->tts_tupleDescriptor,
+													 uncompressed_slot->tts_values,
+													 uncompressed_slot->tts_isnull);
```
```suggestion
                                bool should_free = false;
				uncompressed_tuple = ExecFetchSlotHeapTuple(uncompressed_slot, false, &should_free); 
```

erimatnor commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1449 @@
+			compressed_tuple_found = false;
+			if (num_segkeys)
+			{
+				HeapKeyTest(compressed_tuple,
+							RelationGetDescr(decompressor->in_rel),
+							num_segkeys,
+							segment_by_keys,
+							segment_exists);
+				HeapKeyTest(compressed_tuple,
+							RelationGetDescr(decompressor->in_rel),
+							num_scankeys,
+							segment_order_by_keys,
+							compressed_tuple_found);
```
So, I think what is happening here is that you are testing the uncompressed tuple against the _previously_ found compressed tuple to see if this uncompressed tuple also matches the segment, just like the previously processed uncompressed tuple. 

If this is true, it took me some time to figure this out (and I am still not sure). 

I recommend, instead of having all this code in a _big_ loop, move some of the code into aptly named functions that makes it easier to read and follow the code. For example, here it could make sense to have something like (pseudo code):

```c
HeapTuple compressed_segment_tuple = NULL;

while (uncompressed_tuple = read_next_uncompressed_tuple())
{
    bool tuple_matches_compressed_segment = false;

    while (!tuple_matches_compressed_segment)
    {
        tuple_matches_compressed_segment = tuple_matches_current_compressed_segment(uncompressed_tuple, compressed_segment_tuple)
        if (tuple_matches_compressed_segment) 
        {
              // process uncompressed tuple for matching segment
        }
        else 
        {
            if (compressed_segment_tuple)
                heap_freetuple(compressed_segment_tuple);

            compressed_segment_tuple = read_next_compressed_segment_tuple();
        }
    }
}

erimatnor reviewed (commented):
This was a pretty big and complicated PR, so I will submit a partial review in order to avoid sitting on the review for too long. It will give you a chance to start addressing some of the stuff I've found so far. You can also choose to wait until I complete a full review if you want to avoid having to fix things twice.

Apart from my inline comments, I think you should try to see if you can apply the following:

- Check if you can make function parameters const. It seems many parameters can be const, but are often not.
- I've also seen cases where some function parameters are not used. Check for such cases.
- Take another look at memory usage. I found some cases where it seems memory is leaked. I've pointed out the places I found so far. But generally, I think it makes sense to use per-tuple memory contexts when running scan loops over big tables to ensure memory is freed in each iteration of the loop. I see no use of per-tuple memory context so leaks might easily happen unintentionally.
- Go through codecov and make sure all major code paths are covered. I've seen some entire functions (like the context resize) which is never executed in tests. That code could be broken.

akuzm commented on tsl/src/compression/compression.c:
```diff
@@ -1000,7 +1004,9 @@ row_compressor_process_ordered_slot(RowCompressor *row_compressor, TupleTableSlo
 		row_compressor->first_iteration = false;
 	}
 	bool changed_groups = row_compressor_new_row_is_in_new_group(row_compressor, slot);
-	bool compressed_row_is_full =
+	bool compressed_row_is_full = false;
+
+	compressed_row_is_full =
```
Looks like accidental change.

akuzm commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1457 @@
+			pfree(segment_by_keys);
+		if (segment_order_by_keys)
+			pfree(segment_order_by_keys);
+		/* build scan keys to do lookup into uncompressed chunks */
+		segment_by_keys = build_segment_order_by_scankeys(decompressor,
+														  uncompressed_chunk,
+														  uncompressed_slot,
+														  &null_columns,
+														  &num_segkeys,
+														  SEGMENTBY_KEYS);
+		segment_order_by_keys = build_segment_order_by_scankeys(decompressor,
+																uncompressed_chunk,
+																uncompressed_slot,
+																NULL,
+																&num_scankeys,
+																SEGMENTBY_KEYS | ORDERBY_KEYS);
```
Doing this for each uncompressed rows is a lot of work, you can update the values in ScanKey structures directly instead.

akuzm commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1457 @@
+	ExecDropSingleTupleTableSlot(compressed_slot);
+}
+
+static void
+do_legacy_recompression(RowDecompressor *decompressor, RowCompressor *row_compressor,
+						const CompressTuplesortstateCxt *uncompressed_tuplestorecxt)
+{
```
Let's call it `do_full_recompression` or something like this, to better reflect what it does.

akuzm commented on tsl/src/compression/recompression.c:
```diff
@@ -0,0 +1,1460 @@
+			affectedsegments->num_non_matching_compressedrows[count]--;
+		}
+		tuplesort_performsort(tuplestorecxt->tuplestore);
+		recompress_segment(tuplestorecxt->tuplestore, decompressor->out_rel, row_compressor);
```
Let's make our custom copy of `row_compressor_process_ordered_slot` here? How this would help us:

1) We won't need the `CompressionType` parameter. In general I think it's not a good interface, passing a flag to a function that says "do this" or "do that". Instead we should have different functions. But here, the purpose of the `CompressionType` flag is just to pass it down to the row compressor, so that it can assign the proper sequence numbers. Instead of that, the recompression code should take control of sequence numbers. We know exactly which rows we are going to compress and what sequence numbers the batches should have. So the old interface of "row compressor" that just works row-by-row and assigns sequence numbers itself is not suitable here.

2) `row_compressor_process_ordered_slot` does some extra work inside like `row_compressor_update_group`, but we don't need it here, we know that the rows we are recompressing belong to the same segments. So this is another way in which the old row compressor interface is not appropriate here, and leads to extra work.

mkindahl reviewed (commented):
It is a little unclear what problem this pull request is trying to solve, so having benchmarks showing what improves would be a good argument for this PR. It looks like this optimization would avoid recompressing an entire chunk and improve performance, but given the lack of benchmarks we can not really say if that is the case.

In addition, I think that the procedure is being made more complicated than necessary because you're fetching all the compressed rows into memory. In addition to adding the problem of how to manage the memory, you need to handle different edge cases, which complicates the implementation and generate duplicate code in different branches. Instead of fetching all the compressed rows into memory, I think you can roll the uncompressed rows into the compressed row directly. Something like this (this is a tweaked version of the procedure you've already written):

1. Store all the uncompressed rows into a tuplestore
2. Sort the tuplestore based on the segment-by and order-by columns. You will now have streaks matching the order-by and segment-by keys of the compressed rows.
3. While there are rows in the tuplestore:
   1. Take the first row from the tuple store. This will have segment-by keys and order-by keys.
   2. Find the position of the first compressed row that matches the segment-by keys and is where the order-by key is either before or inside the compressed row.
   3. If this is inside the compressed row you need to roll the uncompressed row (and maybe some more uncompressed rows) into the segment.
	  1. Decompress the row and store it in the tuple store
	  2. Sort the tuplestore
	  3. You now have a sequence of rows mixed from the uncompressed rows and the compressed rows. 
   4. Create a compressed row from the first 1000 rows (or less) in the tuplestore. Here you use the segment-by key and a row counter to decide when to stop reading. This can be done in bounded memory, but you need to switch memory context for this part.
	  - If the compressed row was not full you will now have a new uncompressed row first in the tuplestore or you have run out of uncompressed rows. If you have more rows, iterate to incorporate these, otherwise the loop will exit.
	  - If the compressed row was full you will now have some rows remaining from the original compressed row that needs to be rolled into a compressed row in the segment and it will be done in the next iteration of the loop.

I think this is roughly what @erimatnor and @antekresic have already suggested.

You have most of the pieces already, but if you do it this way, there are a few advantages:
1. You will avoid having to manage the memory allocation for the compressed rows.
2. It will automatically "degenerate" to a full decompression and compression without having to do any threshold checks.

mkindahl commented:
> > If this is inside the compressed row you need to roll the uncompressed row (and maybe some more uncompressed rows) into the segment.
> > ```
> > 1. Decompress the row and store it in the tuple store
> > 
> > 2. Sort the tuplestore
> > 
> > 3. You now have a sequence of rows mixed from the uncompressed rows and the compressed rows.
> > ```
> 
> @mkindahl re-sorting the tuplestore per compressed segment won't do, we had performance problems with it in the previous version. Last time we planned to go with two tuplestores (input rows in one, the rows that we had to re-compress in another), and a sorted merge of them.

Good to know. I was considering this as an option, but assumed that the tuplestore was smart about this. Seems to not be the case.

> Other than that, it's along the lines of what you described. For the reference, the result of the previous discussion with Bharathy, Ante, Sven and myself is here: [#5867 (comment)](https://github.com/timescale/timescaledb/pull/5867#discussion_r1281706453)

Thanks. Will take a look.

sb230132 commented:
> It is a little unclear what problem this pull request is trying to solve, so having benchmarks showing what improves would be a good argument for this PR. It looks like this optimization would avoid recompressing an entire chunk and improve performance, but given the lack of benchmarks we can not really say if that is the case.
> 
> In addition, I think that the procedure is being made more complicated than necessary because you're fetching all the compressed rows into memory. In addition to adding the problem of how to manage the memory, you need to handle different edge cases, which complicates the implementation and generate duplicate code in different branches. Instead of fetching all the compressed rows into memory, I think you can roll the uncompressed rows into the compressed row directly. Something like this (this is a tweaked version of the procedure you've already written):
> 
> 1. Store all the uncompressed rows into a tuplestore
> 2. Sort the tuplestore based on the segment-by and order-by columns. You will now have streaks matching the order-by and segment-by keys of the compressed rows.
> 3. While there are rows in the tuplestore:
>    
>    1. Take the first row from the tuple store. This will have segment-by keys and order-by keys.
>    2. Find the position of the first compressed row that matches the segment-by keys and is where the order-by key is either before or inside the compressed row.
>    3. If this is inside the compressed row you need to roll the uncompressed row (and maybe some more uncompressed rows) into the segment.
>       
>       1. Decompress the row and store it in the tuple store
>       2. Sort the tuplestore
>       3. You now have a sequence of rows mixed from the uncompressed rows and the compressed rows.
>    4. Create a compressed row from the first 1000 rows (or less) in the tuplestore. Here you use the segment-by key and a row counter to decide when to stop reading. This can be done in bounded memory, but you need to switch memory context for this part.
>       
>       * If the compressed row was not full you will now have a new uncompressed row first in the tuplestore or you have run out of uncompressed rows. If you have more rows, iterate to incorporate these, otherwise the loop will exit.
>       * If the compressed row was full you will now have some rows remaining from the original compressed row that needs to be rolled into a compressed row in the segment and it will be done in the next iteration of the loop.
> 
> I think this is roughly what @erimatnor and @antekresic have already suggested.
> 
> You have most of the pieces already, but if you do it this way, there are a few advantages:
> 
> 1. You will avoid having to manage the memory allocation for the compressed rows.
> 2. It will automatically "degenerate" to a full decompression and compression without having to do any threshold checks.

Here are the benchmark results.
https://docs.google.com/spreadsheets/d/1A1OmCfD3mP_cTNqar9YQ20T9puY6xAoLJColYRKKCIs/edit?usp=sharing

horzsolt commented:
We won't continue the work on this one, we will come back to the original problem from a different perspective.
